src/nvim/path.c

```diff
@@ -1650,29 +1646,38 @@ bool vim_isAbsName(char_u *name)
 /// Save absolute file name to "buf[len]".
 ///
 /// @param      fname is the filename to evaluate
-/// @param[out] buf   is the buffer for returning the absolute path for `fname`
+/// @param[out] buf   is the buffer for returning:
+///                     - `fname` truncated if longer than `len`
+///                     - `fname` if `fname` is a URL
+///                     - the absolute path for `fname`
+///                     - `fname` if getting the absolute path fails
 /// @param      len   is the length of `buf`
 /// @param      force is a flag to force expanding even if the path is absolute
 ///
 /// @return           FAIL for failure, OK otherwise
 int vim_FullName(const char *fname, char *buf, size_t len, bool force)
   FUNC_ATTR_NONNULL_ARG(2)
 {
-  int retval = OK;
-  int url;
-
   *buf = NUL;
   if (fname == NULL)
     return FAIL;
 
-  url = path_with_url(fname);
-  if (!url)
-    retval = path_get_absolute_path((char_u *)fname, (char_u *)buf, len, force);
-  if (url || retval == FAIL) {
-    /* something failed; use the file name (truncate when too long) */
+  if (strlen(fname) > (len - 1)) {
+    /* use the truncated file name */
     xstrlcpy(buf, fname, len);
+    return FAIL;
   }
-  return retval;
+
+  if (path_with_url(fname)) {
+    strcpy(buf, fname);
+  } else {
+    if (path_get_absolute_path((char_u *)fname, (char_u *)buf, len, force) == FAIL) {
```

Comment:
`path_get_absolute_path()` may fail for other reasons. E.g. `append_path()` could grow the result beyond `len`. We must keep the behavior the same as before, unless there is a reason to change it.